Scrape to archive

ondenman opened the pull request:
An election has recently been held for this legislature ([29/10/2016](http://www.ipu.org/parline-e/Modlist.asp)).

The data is currently scraped from Wikipedia but the Wikipedia pages have not been updated for the new term. This PR uses http://www.althingi.is as the source for the new term.

The purpose of this PR is to visit each member page belonging to members listed on the Congressmen, Substitutes and Alternates pages so that each page is saved to the archive.

## [Scraper Change checklist](https://github.com/everypolitician/everypolitician/wiki/Scraper-Change-checklist)
* [x] 1. scraper is on Morph.io under the "everypolitician-scrapers" group?
* [x] 2. scraper's GitHub "Website" link points at morph.io page?
* [x] 3. scraper is set to auto-run?
* [x] 4. scraper is archiving?
* [x] 5. legislature has a scraper webhook set? #

ondenman committed:
Add gemfile

ondenman committed:
Bundle install

ondenman committed:
Add MembersPage

Represent a page listing members of the legislature. Exposes a list
of member urls.

ondenman committed:
Archive member pages

ondenman committed:
Add IcelandicMembersPage

Represents a page on which members are listed.

ondenman committed:
Archive Icelandic member pages from Members, Alternates and Substitutes pages

The Icelandic version of the site covers alternates and substitutes
as well as the members listed on the English page.

tmtmtmtm commented on scraper.rb:
```diff
@@ -0,0 +1,20 @@
+
+require_rel 'lib'
+
+# url = 'http://www.althingi.is/altext/cv/en/'
```
not sure why this line is here…

tmtmtmtm commented on scraper.rb:
```diff
@@ -0,0 +1,20 @@
+# url = 'http://www.althingi.is/altext/cv/en/'
+urls = ['http://www.althingi.is/thingmenn/althingismenn/', # members
+        'http://www.althingi.is/thingmenn/thingmenn/varamenn-sem-sitja-a-althingi/', # alternates
+        'http://www.althingi.is/thingmenn/thingmenn/varamenn-sem-hafa-tekid-saeti/'] # substitues
```
Rather than comments, perhaps make this a hash, and then iterate over the values?

tmtmtmtm commented on scraper.rb:
```diff
@@ -0,0 +1,20 @@
+  IcelandicMembersPage.new(response: Scraped::Request.new(url: url).response)
+             .member_urls
+             .each do |url|
+               puts "Opening: #{url}"
```
Too much output like this makes morph very unhappy

tmtmtmtm commented on scraper.rb:
```diff
@@ -0,0 +1,20 @@
+             .member_urls
+             .each do |url|
+               puts "Opening: #{url}"
+               Scraped::HTML.new(response: Scraped::Request.new(url: url).response)
```
might be better to assign this to a nil variable (i.e. `_ = Scraped::HTML.new…`) so that it's more obvious that we're simply ignoring the results for now, rather it being a command method.

tmtmtmtm commented on lib/members_page.rb:
```diff
@@ -0,0 +1,9 @@
+require 'scraped'
+
+class MembersPage < Scraped::HTML
```
is this class being used?

ondenman committed:
Rename MembersPage to EnglishMembersPage

Since we now have IcelandicMembersPage, I have renamed MembersPage
to ENglishMembersPage to prevent any confusion that this might
be a base class. This class specifically represents a page listing
the members of the legislature in English.

ondenman committed:
Store Icelandic members page urls in hash

Replaces need for comments.

ondenman committed:
Scrape both English and Icelandic pages

ondenman committed:
Assign Scraped::HTML object to underscore variable

So that it is clear that we are not actually using the result at
the time being.

ondenman committed:
Remove print statement

tmtmtmtm commented on scraper.rb:
```diff
@@ -0,0 +1,31 @@
+        members: 'http://www.althingi.is/thingmenn/althingismenn/',
+        alternates: 'http://www.althingi.is/thingmenn/thingmenn/varamenn-sem-sitja-a-althingi/',
+        substitues: 'http://www.althingi.is/thingmenn/thingmenn/varamenn-sem-hafa-tekid-saeti/'
+        }
```
This is very odd formatting. As this is a new gem, you should really be making sure it passes `rubocop` checking…

tmtmtmtm commented on scraper.rb:
```diff
@@ -0,0 +1,31 @@
+             .each do |url|
+               scraped_page_for(url)
+             end
+end
```
There's too much duplication here between these sections. I'd suggest harmonising the datastructures, and keying each on the classname you're going to instantiate, so you can collapse this down.

tmtmtmtm commented on scraper.rb:
```diff
@@ -0,0 +1,31 @@
+        }
+
+def scraped_page_for(url)
+  _ = Scraped::HTML.new(response: Scraped::Request.new(url: url).response)
```
This isn't really where you want to do the blank assignment. Now the call to _this_ looks like a command…

ondenman committed:
fixup! Assign Scraped::HTML object to underscore variable

ondenman committed:
fixup! Store Icelandic members page urls in hash

ondenman committed:
Loop over URLs in single structure

This enables us to collapse the two loops for EnglishMembersPage
and IcelandicMembersPage into one.

tmtmtmtm commented on scraper.rb:
```diff
@@ -0,0 +1,29 @@
+
+urls.each do |key, urls|
+  urls.each do |url|
+    eval(key.to_s).new(response: Scraped::Request.new(url: url).response)
```
You can get rid of the `eval` here by making the keys of the data structure be the class names, rather than symbols (i.e `EnglishMembersPage => [ … ]`)

tmtmtmtm commented on scraper.rb:
```diff
@@ -0,0 +1,29 @@
+
+def scraped_page_for(url)
+  Scraped::HTML.new(response: Scraped::Request.new(url: url).response)
+end
```
this is only called from a single place at the minute, and doesn't really add much. You should probably either just inline it, or make it work with the other very similar call too (by also taking a `klass` argument)

tmtmtmtm reviewed (commented):
NB: `rubocop` complains about lots of things in this. I'd suggest setting that to run automatically…

ondenman committed:
Key by classname

Keys the data by classname rather tan symbol therefore negating
the need for eval().

ondenman committed:
Add klass argument to scraped_page_for()

ondenman committed:
Fix formatting for Rubocop

ondenman committed:
Rename variables for Rubocop

Rubocop was complaining that variables were masking outer scope
variables with the same name.

tmtmtmtm reviewed (changes requested):
This code doesn't run:

```
scraper.rb:22:in `block in <main>': undefined local variable or method `urls' for main:Object (NameError)
	from scraper.rb:21:in `each'
	from scraper.rb:21:in `<main>'
```

I would suggest adding a very basic test that ensures that each page has the expected number of URLs, and includes one of the expected URLs.

ondenman committed:
fixup! Rename variables for Rubocop

ondenman commented:
Closing in favour of: https://github.com/everypolitician-scrapers/iceland-althingi/pull/4
